### PR TITLE
Make test-client generic over runtime

### PR DESCRIPTION
**Pull request by tomusdrw**
This allows us to use `generic-test-client` with actual `node/runtime` (see example in `node/executor` tests).

The backward compatibility with `test-client` + `test-runtime` is maintained by extension trait. I imagine that we could build something similar for `test-client` + `node-runtime` if necessary.

@bkchr I've changed couple of builder method names you introduced in your last PR, let me know if you're ok with that.

### CODE REVIEW

File: core/test-client/generic/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+[dependencies]
+client = { package = "substrate-client", path = "../../client" }
+client-db = { package = "substrate-client-db", path = "../../client/db", features = ["test-helpers"] }
+consensus = { package = "substrate-consensus-common", path = "../../consensus/common" }
+executor = { package = "substrate-executor", path = "../../executor" }
+futures = { version = "0.1.17" }
+hash-db = "0.12"
+keyring = { package = "substrate-keyring", path = "../../../core/keyring" }
+parity-codec = "3.5.1"
```
[the remark this answers is not in the document]
Comment (bkchr) [reply core/test-client/generic/Cargo.toml]: No, don't do this. Update will happen in another pr.

File: core/test-client/generic/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+		Executor: client::CallExecutor<Block, Blake2Hasher>,
+		Backend: client::backend::Backend<Block, Blake2Hasher>,
+		Block: BlockT<Hash=<Blake2Hasher as Hasher>::Out>,
+	{
+
+		let storage = {
+			let mut storage = self.genesis_init.genesis_storage();
+
+			// Add some child storage keys.
+			storage.1.insert(
```
Comment (bkchr) [inline core/test-client/generic/src/lib.rs]: Hmm, I think we should make this configurable as well?

File: core/test-client/generic/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+		};
+
+		let client = client::Client::new(
+			self.backend.clone(),
+			executor,
+			storage,
+			self.execution_strategies
+		).expect("Creates new client");
+
+		#[allow(deprecated)]
```
Comment (bkchr) [inline core/test-client/generic/src/lib.rs]: This should not be required anymore.

File: node/executor/Cargo.toml [hunk 4]
```diff
@@ -17,6 +17,7 @@ node-primitives = { path = "../primitives" }
 node-runtime = { path = "../runtime" }
 
 [dev-dependencies]
+test-client = { package = "substrate-generic-test-client", path = "../../core/test-client/generic/" }
```
Comment (Kianenigma) [inline node/executor/Cargo.toml]: Great work and I am already using this. One nit: this is confusing. `test-client` should always point to `path = "../../core/test-client/` (still exists all around the repo) and this should be named `generic-test-client`.

```suggestion
generic-test-client = { package = "substrate-generic-test-client", path = "../../core/test-client/generic/" }
```

File: node/runtime/src/lib.rs [hunk 5]
```diff
@@ -59,7 +59,7 @@ pub const VERSION: RuntimeVersion = RuntimeVersion {
 	impl_name: create_runtime_str!("substrate-node"),
 	authoring_version: 10,
 	spec_version: 92,
-	impl_version: 93,
+	impl_version: 94,
```
Comment (bkchr) [inline node/runtime/src/lib.rs]: As we don't change anything in the runtime, we don't need to bump the `impl_version`.

File: node/executor/src/lib.rs [hunk 6]
```diff
@@ -895,6 +895,21 @@ mod tests {
 		assert!(t.storage_changes_root(GENESIS_HASH.into()).unwrap().is_some());
 	}
 
+	#[test]
+	fn should_import_block_with_test_client() {
```
Comment (Kianenigma) [inline node/executor/src/lib.rs]: Do we currently support the opposite path? The real client using the real runtime? (recalling that this idea came along while I was trying to add a test with the real runtime in `core/basic-authorship`). I had a failed attempt so far.

Comment (bkchr) [reply node/executor/src/lib.rs]: By using the new `generic-test-client` you should be able to do this.

Comment (tomusdrw) [reply node/executor/src/lib.rs]: Yeah, despite the name `test-client` is actually The Real Client™ but with in-memory backend and (previously) test runtime. Now it's configurable and it's just a set of convenience methods to mock the client state.

**Comment by tomusdrw** [comment]
Introduced `test-runtime-client` and moved it to `test-runtime/client`. The `core/test-client` is now the generic one.

**Pull request merged**